File: android/packages/components/src/main/kotlin/io/ente/components/Rows.kt
package io.ente.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

@Composable
public fun InfoRow(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    leading: @Composable () -> Unit,
) {
    val palette = LocalEntePalette.current
    Row(
        modifier
            .fillMaxWidth()
            .heightIn(min = 54.dp)
            .padding(horizontal = EnteSpacing.md, vertical = 9.dp)
            .semantics(mergeDescendants = true) {},
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CompositionLocalProvider(LocalContentColor provides palette.mutedText) {
            Box(Modifier.size(36.dp), contentAlignment = Alignment.Center) { leading() }
        }
        Spacer(Modifier.width(EnteSpacing.md))
        Column(Modifier.weight(1f)) {
            Text(
                title,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
                style = EnteTypography.body,
                color = palette.text,
            )
            if (!subtitle.isNullOrEmpty()) {
                Spacer(Modifier.height(EnteSpacing.xs))
                Text(
                    subtitle,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    style = EnteTypography.mini,
                    color = palette.mutedText,
                )
            }
        }
    }
}

@Composable
public fun MenuGroup(modifier: Modifier = Modifier, content: @Composable () -> Unit) {
    val palette = LocalEntePalette.current
    Surface(
        modifier = modifier.fillMaxWidth(),
        color = palette.surface,
        contentColor = palette.text,
        shape = RoundedCornerShape(EnteRadius.button),
    ) {
        Column { content() }
    }
}

@Composable
public fun PropertyRow(label: String, value: String, modifier: Modifier = Modifier) {
    val palette = LocalEntePalette.current
    Row(
        modifier.fillMaxWidth().heightIn(min = 48.dp).padding(EnteSpacing.md).semantics(
            mergeDescendants = true
        ) {},
        horizontalArrangement = Arrangement.spacedBy(EnteSpacing.md),
    ) {
        Text(
            label,
            Modifier.weight(1f).alignByBaseline(),
            style = EnteTypography.mini,
            color = palette.mutedText,
        )
        Text(
            value,
            Modifier.weight(1f).alignByBaseline(),
            textAlign = TextAlign.End,
            style = EnteTypography.body,
            color = palette.text,
        )
    }
}

@Composable
public fun EnteDivider(modifier: Modifier = Modifier) {
    HorizontalDivider(modifier, color = LocalEntePalette.current.faintBorder)
}

File: apple/packages/EnteComponents/Sources/EnteComponents/Rows.swift
import SwiftUI

public struct InfoRow<Leading: View>: View {
    @Environment(\.entePalette) private var palette
    private let title: String
    private let subtitle: String?
    private let leading: Leading

    public init(_ title: String, subtitle: String? = nil, @ViewBuilder leading: () -> Leading) {
        self.title = title
        self.subtitle = subtitle
        self.leading = leading()
    }

    public var body: some View {
        HStack(spacing: EnteSpacing.md) {
            leading
                .frame(width: 36, height: 36)
                .foregroundStyle(palette.mutedText)
            VStack(alignment: .leading, spacing: EnteSpacing.xs) {
                Text(title)
                    .font(EnteTypography.body)
                    .foregroundStyle(palette.text)
                    .lineLimit(2)
                if let subtitle, !subtitle.isEmpty {
                    Text(subtitle)
                        .font(EnteTypography.mini)
                        .foregroundStyle(palette.mutedText)
                        .lineLimit(2)
                }
            }
            Spacer(minLength: 0)
        }
        .padding(.horizontal, EnteSpacing.md)
        .padding(.vertical, 9)
        .frame(minHeight: 54)
        .accessibilityElement(children: .combine)
    }
}

public struct MenuGroup<Content: View>: View {
    @Environment(\.entePalette) private var palette
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            content
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .foregroundStyle(palette.text)
        .background(palette.surface)
        .clipShape(RoundedRectangle(cornerRadius: EnteRadius.button, style: .continuous))
    }
}

public struct PropertyRow: View {
    @Environment(\.entePalette) private var palette
    private let label: String
    private let value: String

    public init(_ label: String, value: String) {
        self.label = label
        self.value = value
    }

    public var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: EnteSpacing.md) {
            Text(label)
                .font(EnteTypography.mini)
                .foregroundStyle(palette.mutedText)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(value)
                .font(EnteTypography.body)
                .foregroundStyle(palette.text)
                .multilineTextAlignment(.trailing)
                .frame(maxWidth: .infinity, alignment: .trailing)
        }
        .padding(EnteSpacing.md)
        .frame(minHeight: 48)
        .accessibilityElement(children: .combine)
    }
}

public struct EnteDivider: View {
    @Environment(\.entePalette) private var palette

    public init() {}

    public var body: some View {
        palette.faintBorder.frame(height: 1)
    }
}
